contracts/Market.sol

```diff
@@ -1547,23 +1264,66 @@ contract Market is
    * @dev This function is responsible for validating and allocating the supply to fulfill an order.
    * @param purchaser The address of the purchaser.
    * @param certificateAmount The total amount for the certificate.
-   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
-   * @return ids An array of the removal IDs being drawn from to fulfill this order.
-   * @return amounts An array of amounts being allocated from each corresponding removal token.
-   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   * @param supplier The only supplier address from which to purchase carbon removals in this transaction,
+   * or zero address if any supplier is valid.
+   * @param vintages A set of valid vintages from which to allocate removals, empty if any vintage is valid.
+   * @return SupplyAllocationData The removals, amounts, suppliers and count data returned
+   * from the supply allocation algorithm.
+   */
+  function _allocateRemovalsSpecialOrder(
+    address purchaser,
+    uint256 certificateAmount,
+    address supplier,
+    uint256[] memory vintages
+  ) internal returns (SupplyAllocationData memory) {
+    SupplyAllocationData memory supplyAllocationData;
+    if (vintages.length == 0) {
+      // case no restrictions on fulfillment
+      if (supplier == address(0)) {
+        uint256 availableSupply = _removal.getMarketBalance();
+        _validateSupply({
+          certificateAmount: certificateAmount,
+          availableSupply: availableSupply
+        });
+        supplyAllocationData = _allocateSupply({amount: certificateAmount});
+        // case supplier-specific fulfillment only
+      } else {
+        supplyAllocationData = _allocateSupplySingleSupplier({
+          certificateAmount: certificateAmount,
+          supplier: supplier
+        });
+      }
+    } else {
+      // case vintage-specific fulfillment only
+      if (supplier == address(0)) {
+        supplyAllocationData = _allocateSupplySpecificVintages({
+          amount: certificateAmount,
+          vintages: vintages
+        });
+        // case vintage-specific fulfillment and supplier-specific fulfillment
```

Comment:
Consider moving these inside the blocks they relate to instead of at the end of the previous unrelated block